### PR TITLE
Add the substate, actions, and the query component for querying ticket support configuration

### PR DESCRIPTION
**Pull request by southp**
## Summary
This PR adds the related substate and actions for querying ticket support configuration of the current user. Note that it shouldn't be deployed before the `GET /help/tickets/kayako/mine` endpoint code is deployed.

Tasks:

- [x] A new `lib/undocumented.js` function to interact with `GET /help/tickets/kayako/mine`
- [x] Redux settings ( state, actions, reducers ) for encapsulating the data flow
- [x] A query component for the querying actions

## Test Plan
Run <strike>`npm run test-client client/state/ticket-support/`</strike>`npm run test-client client/state/help/ticket/test`

### CODE REVIEW

**Comment by southp** [comment]
Now that the test suites are completed, it should be ready for review. cc @dllh @jordwest @omarjackman .

While the new endpoint it needs has not been deployed yet, running the tests shouldn't have problems.

**Comment by southp** [comment]
Add `Needs i18n Review` label, since I am worrying if the newly-added error message in this PR is clear enough both for plain English and for translation.

File: client/state/notices/middleware.js [hunk 1]
```diff
@@ -143,6 +144,15 @@ export const onPublicizeConnectionUpdateFailure = ( dispatch, { error } ) => dis
 	} ) )
 );
 
+// TODO:
+// Ideally, this error should not be shown if the live chat support is configured properly.
+// However, at the moment the olark-related data is not fully in the redux store, so there
+// is no clean way to access it in this middleware. Once it is in, we should rewrite this function
+// to take that into account.
+const onTicketSupportConfigurationRequestFailure = dispatchError(
+	translate( 'There is an error while configuring the email support form.' )
```
Comment (jordwest) [inline client/state/notices/middleware.js]: This seems like it could be a weird/confusing experience for users who are eligible for live chat, and see this notice. Since it doesn't happen in response to a user action, it might be better handled somewhere other than a global notice anyway. Maybe a notice component at the top of the contact form? That form also has access to the old olark store and could use it to decide whether to display the notice.

File: client/state/ticket-support/configuration/selectors.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+export const defaultConfiguration = {
+	isUserEligible: false,
+};
+
+export const getTicketSupportConfiguration = ( state ) => {
+	return state.ticketSupport.configuration || defaultConfiguration;
+};
```
Comment (jordwest) [inline client/state/ticket-support/configuration/selectors.js]: Other than tests, is there a situation where the entire configuration object would be needed? By using `combineReducers` as mentioned in another comment, we could do away with the `defaultConfiguration` and access the state object directly in tests.

For the individual config options (`isUserEligible` etc), it's nice to have them behind selectors which can easily be updated in one place if the configuration params change.

File: client/components/data/query-ticket-support-configuration/index.jsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * Internal dependencies
+ */
+import {
+	ticketSupportConfigurationRequest,
+} from 'state/ticket-support/configuration/actions';
+
+class QueryTicketSupportConfiguration extends Component {
+	componentWillMount() {
+		this.props.ticketSupportConfigurationRequest();
+	}
```
Comment (jordwest) [inline client/components/data/query-ticket-support-configuration/index.jsx]: It might be useful to check `isRequesting` here to avoid doubling up on requests.

For example, if the `<QueryTicketSupportConfiguration />` component is used in two places at once, they'll both fire off the same request when its only necessary once. Using an `if( ! isRequesting )` check here can avoid that.

File: client/state/action-types.js [hunk 4]
```diff
@@ -498,6 +498,9 @@ export const THEME_CLEAR_ACTIVATED = 'THEME_CLEAR_ACTIVATED';
 export const THEME_DETAILS_REQUEST = 'THEME_DETAILS_REQUEST';
 export const THEME_DETAILS_RECEIVE = 'THEME_DETAILS_RECEIVE';
 export const THEME_DETAILS_RECEIVE_FAILURE = 'THEME_DETAILS_RECEIVE_FAILURE';
+export const TICKET_SUPPORT_CONFIGURATION_REQUEST = 'TICKET_SUPPORT_CONFIGURATION_REQUEST';
+export const TICKET_SUPPORT_CONFIGURATION_REQUEST_SUCCESS = 'TICKET_SUPPORT_CONFIGURATION_REQUEST_SUCCESS';
+export const TICKET_SUPPORT_CONFIGURATION_REQUEST_FAILURE = 'TICKET_SUPPORT_CONFIGURATION_REQUEST_FAILURE';
```
Comment (jordwest) [inline client/state/action-types.js]: This is related to my other comment about changing the location in the state tree; how about `HELP_TICKET_CONFIGURATION_*` for actions? That would help us keep more of the help related stuff together.

File: client/state/action-types.js [hunk 5]
```diff
@@ -498,6 +498,9 @@ export const THEME_CLEAR_ACTIVATED = 'THEME_CLEAR_ACTIVATED';
...
 export const ACTIVE_THEME_REQUEST = 'ACTIVE_THEME_REQUEST';
 export const ACTIVE_THEME_REQUEST_SUCCESS = 'ACTIVE_THEME_REQUEST_SUCCESS';
 export const ACTIVE_THEME_REQUEST_FAILURE = 'ACTIVE_THEME_REQUEST_FAILURE';
```
Comment (jordwest) [inline client/state/action-types.js]: Not part of this PR, but just noticed that this isn't alphabetically ordered 😃

File: client/state/ticket-support/is-requesting/selectors.js [hunk 6]
```diff
@@ -0,0 +1,3 @@
+export const isRequestingTicketSupportConfiguration = ( state ) => {
+	return state.ticketSupport.isRequesting;
+};
```
Comment (jordwest) [inline client/state/ticket-support/is-requesting/selectors.js]: This could probably just go inside `configuration/selectors.js`

File: client/state/ticket-support/configuration/actions.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+	return {
+		type: TICKET_SUPPORT_CONFIGURATION_REQUEST_SUCCESS,
+		...configuration,
+	};
+};
+
+export const ticketSupportConfigurationRequestFailure = ( error ) => {
+	return {
+		type: TICKET_SUPPORT_CONFIGURATION_REQUEST_FAILURE,
+		...error,
```
Comment (southp) [inline client/state/ticket-support/configuration/actions.js]: Nice catch. Fixed by 58ba64dc230931efd0532c3b720a6dd25370a4ec

File: client/state/notices/middleware.js (hunk 1 above)
Comment (southp) [inline client/state/notices/middleware.js]: Agree. I have added a new state property `requestError` by b3849c58ffcfb643000d3327e47fcfcdafbc43bd so that I can conditionally display the prompt according to the chat support eligibility in the following PR. How do you think?

**Comment by southp** [comment]
@jordwest The state tree is now refined as
```
help
-> isReady
-> isRequesting
-> isUserEligible
-> requestError
```
Please see the purpose of `requestError` [in the above reply](https://github.com/Automattic/wp-calypso/pull/9499#discussion_r89745219).

Could you help me give it the second round of reviewing? :)

File: client/state/action-types.js [hunk 8]
```diff
@@ -506,9 +513,9 @@ export const THEME_UPLOAD_FAILURE = 'THEME_UPLOAD_FAILURE';
 export const THEME_UPLOAD_PROGRESS = 'THEME_UPLOAD_PROGRESS';
 export const THEME_UPLOAD_START = 'THEME_UPLOAD_START';
 export const THEME_UPLOAD_SUCCESS = 'THEME_UPLOAD_SUCCESS';
-export const ACTIVE_THEME_REQUEST = 'ACTIVE_THEME_REQUEST';
-export const ACTIVE_THEME_REQUEST_SUCCESS = 'ACTIVE_THEME_REQUEST_SUCCESS';
-export const ACTIVE_THEME_REQUEST_FAILURE = 'ACTIVE_THEME_REQUEST_FAILURE';
+export const TICKET_CONFIGURATION_REQUEST = 'TICKET_CONFIGURATION_REQUEST';
+export const TICKET_CONFIGURATION_REQUEST_SUCCESS = 'TICKET_CONFIGURATION_REQUEST_SUCCESS';
+export const TICKET_CONFIGURATION_REQUEST_FAILURE = 'TICKET_CONFIGURATION_REQUEST_FAILURE';
```
Comment (jordwest) [inline client/state/action-types.js]: Are these actions still being used? Are they different to the `HELP_TICKET_*` actions?

File: client/components/data/query-ticket-support-configuration/index.jsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+	render() {
+		return null;
+	}
+}
+
+export default connect(
+	( state ) => ( {
+		isRequestingTicketSupportConfiguration: isRequestingTicketSupportConfiguration( state ),
```
Comment (jordwest) [inline client/components/data/query-ticket-support-configuration/index.jsx]: Just a minor suggestion, feel free to ignore: Since it's already clear from the name of the component that this is a request for ticket support configuration, the prop could just be named `isRequesting`

File: client/state/help/ticket/test/reducer.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+		} );
+	} );
+
+	it( 'should set isRequesting to true', () => {
+		const state = reducer( undefined, {
+			type: HELP_TICKET_CONFIGURATION_REQUEST,
+		} );
+
+		assert.isTrue( state.isRequesting );
+	} );
```
Comment (jordwest) [inline client/state/help/ticket/test/reducer.js]: Certainly doesn't hurt to have more tests, but these simple uses of `createReducer` are pretty much covered by the `createReducer` tests.

File: client/state/help/ticket/reducer.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+	[ HELP_TICKET_CONFIGURATION_REQUEST_FAILURE ]: () => false,
+} );
+
+const isUserEligible = createReducer( false, {
+	[ HELP_TICKET_CONFIGURATION_REQUEST_SUCCESS ]: ( state, { is_user_eligible } ) => is_user_eligible,
+} );
+
+const isReady = createReducer( false, {
+	[ HELP_TICKET_CONFIGURATION_REQUEST_SUCCESS ]: () => true,
+	[ HELP_TICKET_CONFIGURATION_REQUEST_FAILURE ]: () => true,
```
Comment (jordwest) [inline client/state/help/ticket/reducer.js]: Should this condition be here? If the request fails, shouldn't the state stay the same as it was before the request?

File: client/state/help/ticket/reducer.js [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+} );
+
+const requestError = createReducer( false, {
+	[ HELP_TICKET_CONFIGURATION_REQUEST ]: () => false,
+	[ HELP_TICKET_CONFIGURATION_REQUEST_SUCCESS ]: () => false,
+	[ HELP_TICKET_CONFIGURATION_REQUEST_FAILURE ]: () => true,
+	[ HELP_TICKET_CONFIGURATION_DISMISS_ERROR ]: () => false,
+} );
```
Comment (jordwest) [inline client/state/help/ticket/reducer.js]: Could we store the error in here instead? It might be useful for debugging purposes, and you can still use `!= null` checks to determine whether to display the notice.

File: client/state/action-types.js (hunk 8 above)
Comment (southp) [inline client/state/action-types.js]: Oops. That was a rebasing error. Fixed by f91939bfd34932650eb54de9e3bb5459778febbb

File: client/components/data/query-ticket-support-configuration/index.jsx (hunk 9 above)
Comment (southp) [inline client/components/data/query-ticket-support-configuration/index.jsx]: That's a eco-friendly suggestion :)
Fixed by 6bde82af16800d8653acdbbec14e4eaba855cc77

File: client/state/help/ticket/test/reducer.js (hunk 10 above)
Comment (southp) [inline client/state/help/ticket/test/reducer.js]: True, but I am a bit paranoid on test coverage so I'd like to keep it ;)

File: client/state/help/ticket/reducer.js (hunk 12 above)
Comment (southp) [inline client/state/help/ticket/reducer.js]: That's a good idea. Fixed by 59306726dce831aa9a1bee5948bcaac631816e8c

**Comment by southp** [comment]
@jordwest Thanks for the thorough review as always!
The second round of comments are all addressed now. Could you revisit it to see if I am doing it right?

File: client/state/help/ticket/actions.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+	HELP_TICKET_CONFIGURATION_REQUEST,
+	HELP_TICKET_CONFIGURATION_REQUEST_SUCCESS,
+	HELP_TICKET_CONFIGURATION_REQUEST_FAILURE,
+	HELP_TICKET_CONFIGURATION_DISMISS_ERROR,
+} from 'state/action-types';
+
+export const ticketSupportConfigurationRequestSuccess = ( configuration ) => {
+	return {
+		type: HELP_TICKET_CONFIGURATION_REQUEST_SUCCESS,
+		...configuration,
```
Comment (jordwest) [inline client/state/help/ticket/actions.js]: I just noticed that if the API returned a JSON response containing a `type` property, it would override the action type and could theoretically execute any Redux action. It's unlikely to be a security risk, but seems like something we should fix either way.

Comment (southp) [inline client/state/help/ticket/actions.js]: Nice catch! Fixed by 2e0abcb521516d41788f4b8161d6ff1ba9627e99
